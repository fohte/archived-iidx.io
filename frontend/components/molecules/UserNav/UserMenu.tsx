import * as React from 'react'
import { Dropdown } from 'semantic-ui-react'

import { auth } from '@app/lib/firebaseApp'

export interface Props {
  displayName: string
}

const UserMenu: React.SFC<Props> = ({ displayName }) => (
  <Dropdown item text={displayName} pointing>
    <Dropdown.Menu>
      <Dropdown.Item
        onClick={async () => {
          await auth.signOut()
        }}
      >
        Sign out
      </Dropdown.Item>
    </Dropdown.Menu>
  </Dropdown>
)

export default UserMenu
