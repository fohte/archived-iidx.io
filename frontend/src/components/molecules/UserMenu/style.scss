@import 'styles/colors.scss';

.box {
  position: relative;
}

$item-margin: 0.4em;
$arrow-size: 0.6em;

.dropdown {
  position: absolute;
  right: 0;
  left: auto;
  z-index: 100;

  min-width: 130px;
  margin-top: 1em;
  padding: $item-margin 0;

  background-color: $primary-bg-color;
  border: 1px solid $border-color;
  border-radius: 2px;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);

  &::after {
    position: absolute;
    top: -($arrow-size / 2);
    right: 1em;
    z-index: 50;

    display: block;
    width: $arrow-size;
    height: $arrow-size;

    background-color: $primary-bg-color;
    box-shadow: -1px -1px 0 0 $border-color;
    transform: rotate(45deg);

    content: '';
  }
}

.item {
  display: block;
  width: 100%;
  padding: 8px 12px;

  color: $primary-text-color;

  cursor: pointer;

  &:hover {
    background-color: darken($primary-bg-color, 5%);
  }
}

.divider {
  margin: $item-margin 0;

  border-top: 1px solid $border-color;
}

.display-name {
  margin-right: 0.5em;
}

.hide {
  display: none;
}
