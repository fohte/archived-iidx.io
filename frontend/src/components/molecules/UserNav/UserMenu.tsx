import * as React from 'react'
import { Dropdown } from 'semantic-ui-react'

import { auth } from '@app/lib/firebaseApp'
import { Link } from '@app/routes'

export interface Props {
  displayName: string
}

const UserMenu: React.SFC<Props> = ({ displayName }) => (
  <Dropdown item text={displayName} pointing>
    <Dropdown.Menu>
      <Link route={`/@${displayName}`}>
        <Dropdown.Item>Profile</Dropdown.Item>
      </Link>
      <Dropdown.Divider />
      <Link route="/results/new">
        <Dropdown.Item>Register results</Dropdown.Item>
      </Link>
      <Dropdown.Divider />
      <Dropdown.Item
        onClick={async () => {
          await auth.signOut()
        }}
      >
        Sign out
      </Dropdown.Item>
    </Dropdown.Menu>
  </Dropdown>
)

export default UserMenu
