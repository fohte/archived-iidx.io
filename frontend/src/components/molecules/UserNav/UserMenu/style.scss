@import 'styles/colors.scss';

.box {
  position: relative;
}

$item-margin: 0.4em;
$arrow-size: 0.6em;

.dropdown {
  margin-top: 1em;
  padding: $item-margin 0;
  position: absolute;
  background-color: $primary-bg-color;
  left: auto;
  right: 0;
  border-radius: 2px;
  min-width: 130px;
  z-index: 100;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
  border: 1px solid $lighter-gray;

  a {
    outline: none;
    text-decoration: none;
  }

  &:after {
    content: '';
    position: absolute;
    top: -($arrow-size / 2);
    right: 1em;
    display: block;
    transform: rotate(45deg);
    width: $arrow-size;
    height: $arrow-size;
    background-color: $white;
    box-shadow: -1px -1px 0 0 $border-color;
    z-index: 50;
  }
}

.item {
  padding: 8px 12px;
  width: 100%;
  color: $primary-text-color;
  cursor: pointer;

  &:hover {
    background-color: $dark-white;
  }
}

.divider {
  border-top: 1px solid $border-color;
  margin: $item-margin 0;
}

.display-name {
  margin-right: 0.5em;
}
