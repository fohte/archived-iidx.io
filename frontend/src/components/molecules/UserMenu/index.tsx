import { faCaretDown } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import * as classnames from 'classnames/bind'
import * as React from 'react'

import { auth } from '@app/lib/firebaseApp'
import { Link } from '@app/routes'
import * as css from './style.scss'

const cx = classnames.bind(css)

export interface Props {
  displayName: string
  className?: string
}

const UserMenu: React.SFC<Props> = ({ displayName, className }) => {
  const [isOpen, setOpeningState] = React.useState(false)

  const boxRef = React.useRef<HTMLDivElement | null>(null)

  const handleClickOutside = (event: any) => {
    if (isOpen && boxRef.current && !boxRef.current.contains(event.target)) {
      setOpeningState(false)
    }
  }

  React.useEffect(() => {
    document.addEventListener('click', handleClickOutside, true)
    return () => {
      document.removeEventListener('click', handleClickOutside, true)
    }
  })

  const toggleOpeningState = () => {
    setOpeningState(!isOpen)
  }
  const hide = () => {
    if (isOpen) {
      setOpeningState(false)
    }
  }

  return (
    <div className={cx('box')} ref={boxRef}>
      <div className={className} onClick={toggleOpeningState}>
        <span className={cx('displayName')}>{displayName}</span>
        <FontAwesomeIcon icon={faCaretDown} size="xs" />
      </div>
      <div className={cx('dropdown', { hide: !isOpen })}>
        <Link route={`/@${displayName}`}>
          <a className={cx('item')} onClick={hide}>
            Profile
          </a>
        </Link>
        <div className={cx('divider')} />
        <Link route="/results/new">
          <a className={cx('item')} onClick={hide}>
            Register results
          </a>
        </Link>
        <div className={cx('divider')} />
        <div
          className={cx('item')}
          onClick={async () => {
            hide()
            await auth().signOut()
          }}
        >
          Sign out
        </div>
      </div>
    </div>
  )
}

export default UserMenu
