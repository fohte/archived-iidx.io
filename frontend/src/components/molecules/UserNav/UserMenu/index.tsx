import * as React from 'react'

import { auth } from '@app/lib/firebaseApp'
import { Link } from '@app/routes'
import * as css from './style.scss'

export interface Props {
  displayName: string
}

const UserMenu: React.SFC<Props> = ({ displayName }) => (
  <div className={css.box}>
    {displayName}
    <div className={css.dropdown}>
      <Link route={`/@${displayName}`}>
        <a>
          <div className={css.item}>Profile</div>
        </a>
      </Link>
      <div className={css.divider} />
      <Link route="/results/new">
        <a>
          <div className={css.item}>Register results</div>
        </a>
      </Link>
      <div className={css.divider} />
      <div
        className={css.item}
        onClick={async () => {
          await auth.signOut()
        }}
      >
        Sign out
      </div>
    </div>
  </div>
)

export default UserMenu
