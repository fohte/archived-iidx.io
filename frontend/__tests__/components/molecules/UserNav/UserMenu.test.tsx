import * as React from 'react'
import * as renderer from 'react-test-renderer'

import UserMenu from '@app/components/molecules/UserNav/UserMenu'

describe('UserMenu', () => {
  it('renders correctly', () => {
    const component = renderer.create(<UserMenu displayName="foobar" />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
