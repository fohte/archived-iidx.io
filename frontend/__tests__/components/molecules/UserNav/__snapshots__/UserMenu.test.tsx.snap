// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMenu renders correctly 1`] = `
<div
  aria-expanded={false}
  className="ui item pointing dropdown"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  role="listbox"
  tabIndex={0}
>
  <div
    aria-live="polite"
    className="text"
    role="alert"
  >
    foobar
  </div>
  <i
    aria-hidden="true"
    className="dropdown icon"
    onClick={[Function]}
  />
  <div
    className="menu transition"
  >
    <div
      className="item"
      onClick={[Function]}
      role="option"
    >
      Sign out
    </div>
  </div>
</div>
`;
